### PR TITLE
Test zones for the CNAME function in Recursor.pm in Engine

### PR DESCRIPTION
**Pull request by matsduf**
## Purpose

Test zone for unit tests for https://github.com/zonemaster/zonemaster-engine/pull/1288

## How to test this PR

Create unit tests based on these test zones when the PR in engine has been completed.

### CODE REVIEW

File: docs/public/specifications/test-zones/Engine/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+* [Connectivity-TP](Connectivity-TP)
+* [Consistency-TP](Consistency-TP)
+* [DNSSEC-TP](DNSSEC-TP)
+* [Delegation-TP](Delegation-TP)
+* [Nameserver-TP](Nameserver-TP)
+* [Syntax-TP](Syntax-TP)
+* [Zone-TP](Zone-TP)
+
+
+Test zones for the Methods defined in [MethodsV2] are also not found in this structuer. They
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/Engine/README.md:29]:
```suggestion
Test zones for the Methods defined in [MethodsV2] are also not found in this structure. They
```

File: docs/public/specifications/test-zones/Engine/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Specification of test zones for Engine
+
+This document structure contains test zone specifications for Zonemaster-Engine,
+except for test zones for Test Case implementations and implementation of
+MethodV2, which are found in sibling structure.
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/Engine/README.md:5]:
```suggestion
MethodV2, which are found in a sibling structure.
```

File: docs/public/specifications/test-zones/Engine/README.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+The test zones found here are for various Perl modulues. The specifications are
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/Engine/README.md:7]:
```suggestion
The test zones found here are for various Perl modules. The specifications are
```

File: docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+## Test zone name
+
+The test zone for these scenarios is `cname.recursor.engine.xa`. Subdomain and
+subzones are created. The names are given in section
+"[Zone setup for test scenarios]" below.
+
+
+## Test scenarios
+
+Scenario name                | Expectec output
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md:33]:
```suggestion
Scenario name                | Expected output
```

File: docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+   target.goodsub          A     127.0.0.1
+```
+
+### NXDOMAIN-VIA-CNAME
+The query name exists, but as CNAME record. The target name of CNAME does not
+exist.
+
+* Query name: "nxdomain-via-cname.cname.recursor.engine.xa"
+  * The [RCODE Name] in the response is NxDomain.
+  * The target of the CNAME, `nxdomain-via-cname-target` does not exist in the,
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md:130]:
```suggestion
  * The target of the CNAME, `nxdomain-via-cname-target` does not exist in the
```

File: docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    zone and is not delegated.
+  * To be found in the answer section:
+```
+   nxdomain-via-cname       CNAME nxdomain-via-cname-target
+```
+
+### NODATA-VIA-CNAME
+The query name exists, but as CNAME record. The target name of CNAME does not
+exist.
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md:139]: The target name exists as a node, but does not contain any resource records of type A.

File: docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+### MULT-CNAME
+The query name exists, but as CNAME, as two CNAME records.
+
+* Query name: "mult-cname.cname.recursor.engine.xa"
+  * To be found in the answer section:
+```
+   mult-cname            CNAME mult-cname-target-1
+   mult-cname            CNAME mult-cname-target-2
+   mult-cname-target-1   A     127.0.0.1
+   mult-cname-target-2   A     127.0.0.2
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md:158]: I think the response may or may not contain the A records.

After experimenting with cases of illegal uses of CNAME like this one in IBDNS, I came to the conclusion that I can't have IBDNS follow aliases in those cases. In fact, you’d hit problem cases that are similar to the reason why DNS query packets never contain more than one question: what if `mult-cname-target-1` exists but `mult-cname-target-2` doesn’t? What RCODE do you set in the response header?

So I suggest that the answer section should contain the two CNAME resource records, but without expecting the A records. Either way, it’ll be an illegal answer.

**Review by marc-vanderwal — COMMENTED** [review]
Looks fine. I just saw a few typos and I have one comment about a problem case.

**Review by matsduf — COMMENTED** [review]
@marc-vanderwal, also see the `dig` output in `test-zone-data/Engine/Recursor-PM/README.md` in this PR.

File: docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md (hunk 7 above)
Comment (marc-vanderwal) [reply docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md:160]: After sleeping over it, I think you’re right and there is no need to change anything; please disregard my comment.

I was just wondering how an intentionally broken DNS server could systematically respond to zone files containing such illegal aliases when there are multiple CNAMEs for the same owner name. But this is not the right place to discuss this sort of thing.

File: test-zone-data/Engine/Recursor-PM/README.md [hunk 8]
```diff
@@ -20,6 +20,446 @@ See [CNAME.md] for specification of the scenarios.
...
+;; AUTHORITY SECTION:
+sub3.cname.recursor.engine.xa. 3600 IN	NS	ns1.sub3.cname.recursor.engine.xa.
+
+;; Query time: 0 msec
+;; SERVER: 127.30.1.33#53(127.30.1.33) (UDP)
+;; WHEN: Wed Nov 22 21:47:03 UTC 2023
+;; MSG SIZE  rcvd: 293
+```
+
 To be completed.
```
Comment (marc-vanderwal) [inline test-zone-data/Engine/Recursor-PM/README.md:463]: This section is now complete, isn’t it?

File: test-zone-data/Engine/Recursor-PM/README.md [hunk 9]
```diff
@@ -20,6 +20,446 @@ See [CNAME.md] for specification of the scenarios.
 
 ## zonemaster-cli commands and their output for each test scenario
```
[the remark this answers is not in the document]
Comment (matsduf) [reply test-zone-data/Engine/Recursor-PM/README.md:21]: The outputs are with `dig` against the `CoreDNS` configuration. They are there to show what Zonemaster would get if the unit tests query for those names with query type `A`. 

I am waiting for https://github.com/zonemaster/zonemaster-engine/pull/1288 to be done.

File: test-zone-data/Engine/Recursor-PM/README.md (hunk 8 above)
Comment (matsduf) [reply test-zone-data/Engine/Recursor-PM/README.md:463]: I guess we could consider it to be complete. I will remove the "to be completed".

**Comment by matsduf** [comment]
@marc-vanderwal and @tgreenx: Please re-review.

**Review by marc-vanderwal — DISMISSED** [review]

**Comment by matsduf** [comment]
I want to hear from @tgreenx before merging. It should match https://github.com/zonemaster/zonemaster-engine/pull/1288.

File: test-zone-data/Engine/Recursor-PM/README.md [hunk 10]
```diff
@@ -0,0 +1,465 @@
...
+looped-cname-in-zone-1.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-1.cname.recursor.engine.xa.
+looped-cname-in-zone-1.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-1.cname.recursor.engine.xa.
+looped-cname-in-zone-1.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-1.cname.recursor.engine.xa.
+looped-cname-in-zone-1.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-1.cname.recursor.engine.xa.
+looped-cname-in-zone-1.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-1.cname.recursor.engine.xa.
+looped-cname-in-zone-1.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-1.cname.recursor.engine.xa.
+looped-cname-in-zone-1.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-1.cname.recursor.engine.xa.
+looped-cname-in-zone-1.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-1.cname.recursor.engine.xa.
+looped-cname-in-zone-1.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-1.cname.recursor.engine.xa.
+looped-cname-in-zone-1.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-1.cname.recursor.engine.xa.
```
[the remark this answers is not in the document]
Comment (matsduf) [reply test-zone-data/Engine/Recursor-PM/README.md:298]: I guess it is a bug in CoreDNS, but on the other hand, the CNAME records are identical and I feel they should be considered to be one. I will check.

File: test-zone-data/Engine/Recursor-PM/README.md [hunk 11]
```diff
@@ -0,0 +1,465 @@
...
+looped-cname-in-zone-2.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-2-a.cname.recursor.engine.xa.
+looped-cname-in-zone-2-a.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-b.cname.recursor.engine.xa.
+looped-cname-in-zone-2-b.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-a.cname.recursor.engine.xa.
+looped-cname-in-zone-2-a.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-b.cname.recursor.engine.xa.
+looped-cname-in-zone-2-b.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-a.cname.recursor.engine.xa.
+looped-cname-in-zone-2-a.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-b.cname.recursor.engine.xa.
+looped-cname-in-zone-2-b.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-a.cname.recursor.engine.xa.
+looped-cname-in-zone-2-a.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-b.cname.recursor.engine.xa.
+looped-cname-in-zone-2-b.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-a.cname.recursor.engine.xa.
+looped-cname-in-zone-2-a.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-b.cname.recursor.engine.xa.
```
[the remark this answers is not in the document]
Comment (matsduf) [reply test-zone-data/Engine/Recursor-PM/README.md:338]: Same answer as above.

[the remark this answers is not in the document]
Comment (matsduf) [reply test-zone-data/Engine/Recursor-PM/README.md:338]: I have also reported this, but here I found a work-around wht with dig gives the following result:
```
; <<>> DiG 9.18.18-0ubuntu0.22.04.1-Ubuntu <<>> @127.30.1.31 looped-cname-in-zone-2.cname.recursor.engine.xa
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 13110
;; flags: qr aa rd; QUERY: 1, ANSWER: 3, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
; COOKIE: 7a7a75c08fe99560 (echoed)
;; QUESTION SECTION:
;looped-cname-in-zone-2.cname.recursor.engine.xa. IN A

;; ANSWER SECTION:
looped-cname-in-zone-2.cname.recursor.engine.xa. 3600 IN CNAME looped-cname-in-zone-2-a.cname.recursor.engine.xa.
looped-cname-in-zone-2-a.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-b.cname.recursor.engine.xa.
looped-cname-in-zone-2-b.cname.recursor.engine.xa. 3600	IN CNAME looped-cname-in-zone-2-a.cname.recursor.engine.xa.

;; AUTHORITY SECTION:
cname.recursor.engine.xa. 3600	IN	NS	ns1.cname.recursor.engine.xa.

;; Query time: 0 msec
;; SERVER: 127.30.1.31#53(127.30.1.31) (UDP)
;; WHEN: Wed Nov 29 13:07:05 UTC 2023
;; MSG SIZE  rcvd: 488
```

I will update.

**Comment by matsduf** [comment]
@marc-vanderwal and @tgreenx: Please re-review.

**Comment by matsduf** [comment]
@marc-vanderwal and @tgreenx: Please re-review.

File: docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+   too-long-cname-chain-three        CNAME too-long-cname-chain-four
+   too-long-cname-chain-four         CNAME too-long-cname-chain-five
+   too-long-cname-chain-five         CNAME too-long-cname-chain-six
+   too-long-cname-chain-six          CNAME too-long-cname-chain-seven
+   too-long-cname-chain-seven        CNAME too-long-cname-chain-eight
+   too-long-cname-chain-eight        CNAME too-long-cname-chain-nine
+   too-long-cname-chain-nine         CNAME too-long-cname-chain-ten
+   too-long-cname-chain-ten          CNAME too-long-cname-chain-target
+   too-long-cname-chain-target       A     127.0.0.1
+```
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md:269]: Yes, but CoreDNS can only produce 10, and that is already too much. Let us lower the limit to 9 inclusive.

File: docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+  * To be found in the answer section:
+```
+   broken-cname-chain            CNAME broken-cname-chain-two
+   broken-cname-chain-three      CNAME broken-cname-chain-target
+   broken-cname-chain-target     A     127.0.0.1
+```
+
+
+### WRONG-CNAME-OWNER-NAME
+The owner name of the CNAME in the response does not match queryy name.
```
[the remark this answers is not in the document]
Comment (matsduf) [reply docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md]: Fixed.

File: docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+MULT-CNAME                   | Undefined and tag `CNAME_MULTIPLE_FOR_NAME`
+LOOPED-CNAME-IN-ZONE-1       | Undefined and tag `CNAME_LOOP_INNER`
+LOOPED-CNAME-IN-ZONE-2       | Undefined and tag `CNAME_LOOP_INNER`
+LOOPED-CNAME-IN-ZONE-3       | Undefined and tag `CNAME_LOOP_INNER`
+LOOPED-CNAME-OUT-OF-ZONE     | Undefined and tag `CNAME_LOOP_OUTER`
+TOO-LONG-CNAME-CHAIN         | ??
+TARGET-NO-MATCH-CNAME        | ??
+BROKEN-CNAME-CHAIN           | ??
+WRONG-CNAME-OWNER-NAME       | ??
+EXTRA-CNAME-IN-ANSWER        | ??
```
[the remark this answers is not in the document]
Comment (marc-vanderwal) [reply docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md:50]: Are there any tags to be output alongside `WRONG-CNAME-OWNER-NAME` and `EXTRA-CNAME-IN-ANSWER`?

File: docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+MULT-CNAME                   | Undefined and tags `CNAME_START`, `CNAME_MULTIPLE_FOR_NAME`
+LOOPED-CNAME-IN-ZONE-1       | Undefined and tags `CNAME_START`, `CNAME_LOOP_INNER`
+LOOPED-CNAME-IN-ZONE-2       | Undefined and tags `CNAME_START`, `CNAME_LOOP_INNER`
+LOOPED-CNAME-IN-ZONE-3       | Undefined and tags `CNAME_START`, `CNAME_LOOP_INNER`
+LOOPED-CNAME-OUT-OF-ZONE     | Undefined and tags `CNAME_START`, `CNAME_LOOP_OUTER`
+TOO-LONG-CNAME-CHAIN         | Undefined and tags `CNAME_START`, `CNAME_RECORDS_TOO_MANY`
+TARGET-NO-MATCH-CNAME        | Undefined and tags `CNAME_START`, `CNAME_NO_MATCH`
+BROKEN-CNAME-CHAIN           | Undefined and tags `CNAME_START`, `CNAME_RECORDS_CHAIN_BROKEN`
+WRONG-CNAME-OWNER-NAME       | False
+EXTRA-CNAME-IN-ANSWER        | False
```
Comment (marc-vanderwal) [inline docs/public/specifications/test-zones/Engine/Recursor-PM/CNAME.md:50]:
```suggestion
WRONG-CNAME-OWNER-NAME       | False and no tags
EXTRA-CNAME-IN-ANSWER        | False and no tags
```

**Review by marc-vanderwal — APPROVED** [review]

**Pull request merged**